### DISCUSSION TITLE
Make writing workflows easier/faster with pre-computed outputs

### DISCUSSION THREAD

**Discussion by Jiehong** [Netflix/conductor#2361] [Ideas]
Hi,

While writing workflows, we've realised that testing them can be tricky. Especially when a workflow becomes pretty big, testing it means running it again, usually from the very beginning, and it can take a long time.

Some parts of a workflow are not so easy to test (like calling a service that fails, whose failure is not under our control).

Inspired by "mocks" in unit-tests, I'd like to propose something to both reduce the feedback loop of a workflow, and improve the ability to force a workflow to handle hard to reproduce states.

## Proposal

A new optional parameter named "mockOutputParameters" is added to a Task within a Workflow.

In the Decider, if this new parameter is not present, the behaviour stays the same.

If present, the task is directly set to COMPLETED, and its outputData is set to the content of the mockOutputParameters.

Note: perhaps the status should also be easy to mock? Maybe.

This allows to rig a workflow in a such a way that only the few tasks being designed/rigged together are actually executed in the workflow, while leaving the rest the same. Not actually running the tasks considerably improves the speed of execution.

It almost allows a "repl-like experience" to write workflows (well, not quite, but closer).

I'd be willing to implement it if agreed upon.

## What this is not

This is a replacement for the resume and retry at all.
This is only meant as a way to improve the speed at which one can design a new workflow. The mockOutputParameters would be all cleaned up when it is ready and then used as such.

**Comment by apanicker-nflx** [comment]
This is definitely a good addition to Conductor and will greatly improve the usability and reduce the onboarding costs. This could make for a great supplemental addition to the debug mode execution feature that is on our [roadmap](https://github.com/Netflix/conductor/wiki/Roadmap#usability) under _Improved developer experience_ .

The `mockOutputParameters` could be supplied as an input when a workflow is being run in debug mode and in this mode, the state machine can skip over and apply these outputs to the tasks that have been mocked. For tasks that are not mocked, the workflow would pause and wait for user input to progress further. Please let us know your thoughts about this.

**Reply by Jiehong** [reply]
I mostly agree, but:

> For tasks that are not mocked, the workflow would pause and wait for user input to progress further

I'm not quite sure how the user would be providing the input. Do you mean the actual workers would not be presented with the task, and the user could update it with the wanted outputs?

At first, I wanted to just have mockOutputParameters be taken into account, but otherwise execute the workflow as usual.
Might just be a first step in the Improved developer experience, though.

But that's a detail, and we agree on the overall idea and concept.

What is the timeline of the linked roadmap, if any?

**Reply by apanicker-nflx** [reply]
> I'm not quite sure how the user would be providing the input. Do you mean the actual workers would not be presented with the task, and the user could update it with the wanted outputs?

By user input here, what I was trying to convey is a signal from the user to continue execution of the task. This could be a simple button the UI which translates to an API call (or similar). Additionally, this could be extended to allow the user to optionally tweak the input to a given task which is curated by the state machine based on previous task output/workflow input.

>What is the timeline of the linked roadmap, if any?

We do not have a strict timeline for this roadmap. The roadmap serves as a guiding vision for the direction that the project is headed in and scalability and developer experience are some of the highest priority items that we would like to address in the short term.
